util/generate_compilation_db.py

```diff
@@ -92,6 +103,29 @@ def __init__(self, action: Dict):
         self.arguments = action.get('arguments', None)
         self.input_dep_set_ids = action.get('inputDepSetIds', [])
 
+    def filter_arguments_for_clangd(self) -> List[str]:
+        args = self.arguments
+        out = []
+        while len(args) > 0:
+            if args[0] == '-o':
+                args = args[2:]
+            elif args[0] == '-MD':
+                args = args[1:]
+            elif args[0] == '-MF':
+                args = args[2:]
+            elif args[0].startswith('-frandom-seed='):
+                args = args[1:]
+            elif args[0].startswith("-march="):
+                args = args[1:]
+            elif args[0].startswith("-mabi="):
+                args = args[1:]
```

Comment:
What problem did these two options cause? I guess for a generic "does this compile OK" IDE diagnostics test it doesn't matter too much, but not having those options could cause spurious errors in some corner cases, like inline assembly, builtins, etc.

Reply:
Let's see, when I don't strip the `-march` and `-mabi` args, I get a message from clangd saying basically that it failed to launch a compiler instance because I asked for an unsupported CPU.

```
external/crt/toolchains/lowrisc_rv32imcb/wrappers/clang -iquote . -iquote bazel-out/k8-fastbuild/bin -nostdinc -isystemexternal/lowrisc_rv32imcb_files/lib/clang/13.0.1/include -isystemexternal/lowrisc_rv32imcb_files/riscv32-unknown-elf/include -isystemexternal/lowrisc_rv32imcb_files/riscv32-unknown-elf/include/c++/10.2.0 -isystemexternal/lowrisc_rv32imcb_files/riscv32-unknown-elf/include/c++/10.2.0/backward -isystemexternal/lowrisc_rv32imcb_files/riscv32-unknown-elf/include/c++/10.2.0/riscv32-unknown-elf -march=rv32imc -mabi=ilp32 -mcmodel=medany -mlittle-endian -Werror -Werror=date-time -ffunction-sections -fdata-sections -Os -g -ffreestanding -fno-common  -std=gnu11 -fno-builtin -c -resource-dir=/usr/lib/clangd/lib/clang/google3-trunk -- /usr/local/google/home/dmcardle/REPOS/opentitan/sw/device/lib/base/memory.c
I[09:38:23.397] --> textDocument/clangd.fileStatus
E[09:38:23.398] Could not build a preamble for file /usr/local/google/home/dmcardle/REPOS/opentitan/sw/device/examples/sram_program/sram_program.c version 1: CreateTargetInfo() return null
I[09:38:23.398] --> workspace/semanticTokens/refresh(0)
I[09:38:23.398] --> textDocument/clangd.fileStatus
I[09:38:23.398] --> textDocument/clangd.fileStatus
E[09:38:23.398] Failed to prepare a compiler instance: valid target CPU values are: nocona, core2, penryn, bonnell, atom, silvermont, slm, goldmont, goldmont-plus, tremont, nehalem, corei7, westmere, sandybridge, corei7-avx, ivybridge, core-avx-i, haswell, core-avx2, broadwell, skylake, skylake-avx512, skx, cascadelake, cooperlake, cannonlake, icelake-client, rocketlake, icelake-server, tigerlake, sapphirerapids, alderlake, knl, knm, k8, athlon64, athlon-fx, opteron, k8-sse3, athlon64-sse3, opteron-sse3, amdfam10, barcelona, btver1, btver2, bdver1, bdver2, bdver3, bdver4, znver1, znver2, znver3, x86-64, x86-64-v2, x86-64-v3, x86-64-v4
```


Stripping these flags is admittedly hacky, but I'm not sure what the proper solution would be. Perhaps we'd need to compile our own clangd that enables march=rv32imc support.

Reply:
This might be relevant: https://github.com/clangd/clangd/issues/1151

I think at minimum, I think I should add a comment on the lines you've highlighted to explain why we're stripping the flags.

Reply:
Just pushed some changes that simplify the flag stripping logic, add a comment, and fix the tests. PTAL, @luismarques :)

It turned out I was stripping more flags than I needed to. I suspect clangd strips some flags by default. It actually exposes an [option for controlling flag stripping](https://clangd.llvm.org/config.html#compileflags) in `.clangd`, but given that we're generating the file ourselves, we might as well do it here.